### PR TITLE
Feature/multex

### PR DESCRIPTION
**Pull request by kellino**
A large merge into dev, the bulk of which is concerned with multiple explanations for images only. In addition to adding this feature, this pull request also 

- fixes database writing for single explanations , multi explanations and spectral explanations
- fixes writing additional data into the `Prediction` object
- makes insertion/deletion curve normalisation an option in `rex.toml` 
- moves spatial explanation extraction to pure torch and away from numpy
- adds a number of extra options into `rex.toml`
- simplifies cmdline args
- introduces `contrastive` explanation type
- adds helper functions into `database.py` for interactive use
- various other small bug fixes

### CODE REVIEW

**Review by liz-is — COMMENTED** [review]
Looks good, but I have some suggestions! 

After this morning's bug fixes this runs for me now with the starfish - let me know if there's another example image you use for testing the multi-explanations.

I haven't yet tested out the database functions, or explored the plotting options. I'd like to check out those too before merging but I thought I'd share my initial comments now

File: rex_xai/visualisation.py [hunk 1]
```diff
@@ -308,15 +347,44 @@ def save_image(explanation, data: Data, args: CausalArgs, path=None):
...
+            for c in clauses:
+                explanation = explanations[c]
+                explanation = __transpose_mask(explanation, data.mode, data.transposed)
+                if explanation is not None:
+                    if composite_mask is None:
+                        composite_mask = explanation
+                    else:
+                        composite_mask = np.where(explanation, 1, composite_mask)
+
+                if data.mode == "first":
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/visualisation.py:362]: yes!

File: rex_xai/extraction.py [hunk 2]
```diff
@@ -121,73 +105,102 @@ def __global(self, map=None, wipe=False):
...
         cutoff = (
             self.data.model_width * self.data.model_height * self.data.model_channels  # type: ignore
         )
         while tt.count_nonzero(mask) < cutoff:
             if expansion_limit is not None:
-                if expansions >= expansion_limit:
+                if expansions >= expansion_limit and expansion_limit > 1:
                     logger.info(
                         f"no explanation found after {expansion_limit} expansions"
                     )
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/extraction.py:177]: sounds good

File: rex_xai/_utils.py [hunk 3]
```diff
@@ -65,6 +89,8 @@ def get_device(gpu: bool):
 
 
 def get_map_locations(map, reverse=True):
+    if isinstance(map, tt.Tensor):
+        map = map.detach().cpu().numpy()
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/_utils.py:102]: a more logical and consistent approach on my part would probably mean this code isn't needed. As it is, I'm not adverse to having these checks in place, just in case...

File: rex_xai/config.py [hunk 4]
```diff
@@ -503,11 +547,12 @@ def load_config(config_path=None):
         try:
             process_config_dict(config_file_args, default_args)
             return default_args
-        except Exception:
+        except Exception as e:
+            print(f"{e}, reverting to default args")
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/config.py:551]: I'd genuinely forgotten that logger was imported into `config.py`. I'll resolve these

File: rex_xai/config.py [hunk 5]
```diff
@@ -221,6 +226,20 @@ def cmdargs():
...
+        const=10,
+        help="multiple explanations, with optional number <x> of floodlights, defaults to value in rex.toml, or 10 if undefined",
+    )
+
+    parser.add_argument(
+        "--contrastive",
+        nargs="?",
+        const=10,
+        help="a contrastive explanation, both necessary and sufficient, needs optional number <x> of floodlights, defaults to value in rex.toml, or 10 if undefined",
+    )
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/config.py:242]: Yes, I'll fill you in when we next meet. The code isn't finished for that, but it's built on top of multex. It's a bit messy I know to introduce it in this particular pull request, but ah well

**Comment by liz-is** [comment]
The peacock is a nice example for testing with, thanks! The plotting on the command line works nicely too, the only comment I have there is that the differently coloured boundaries are not very visible at the resolution the images are at - can they be made thicker?

**Comment by liz-is** [comment]
I added `mode="thick"` to the `mark_boundaries` calls - this makes the lines 2 pixels wide instead of 1, which helps a lot.

**Comment by liz-is** [comment]
I had a chance to check the new database functionality too and it works nicely! I'll add an example of using this to my draft interactive tutorial in #37

File: rex_xai/visualisation.py [hunk 6]
```diff
@@ -18,6 +19,22 @@
...
+    (1, 1, 1),
+    (1, 0, 1),
+    (0, 1, 1),
+    (1, 1, 0),
+    (0.5, 0.5, 0.5),
+    (0.25, 0.25, 0.25),
+    (0, 0.75, 1),
+]
+
+
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/visualisation.py:37]: I think I have a nice solution to this which isn't bounded.

File: rex_xai/config.py [hunk 7]
```diff
@@ -61,23 +61,28 @@ def __init__(self) -> None:
...
-        self.no_expansions = 50
+        self.no_expansions = 4
         # spotlight args
         self.spotlights: int = 10
         self.spotlight_size: int = 20
         self.spotlight_eta: float = 0.2
         self.spotlight_step: int = 5
-        self.spotlight_objective_function = np.mean
+        self.spotlight_objective_function = None
+        self.spotlight_objective_function_name: str = "none"
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/config.py]: Fixed

File: rex_xai/multi_explanation.py [hunk 8]
```diff
@@ -52,7 +56,14 @@ def extract(self, method=None):
...
             self.blank()
-            for i in range(0, self.args.spotlights):
+            # we start with the global max explanation
+            self._Explanation__global()
+            if self.final_mask is not None:
+                self.explanations.append(self.final_mask)
+
+            self.blank()
+
+            for i in range(0, self.args.spotlights - 1):
```
Comment (liz-is) [inline rex_xai/multi_explanation.py:66]: Wasn't this line correct before? I was getting the expected number of outputs!

File: rex_xai/multi_explanation.py [hunk 9]
```diff
@@ -26,8 +26,10 @@ def save(self, path, mask=None, multi=None, multi_style=None, clauses=None):
         if multi_style == "separate":
             logger.info("saving explanations in multiple different files")
             for i, mask in enumerate(self.explanations):
-                name, ext = os.path.splitext(path)
-                super().save(f"{name}_{i}{ext}", mask=mask)
+                if path is not None:
+                    name, ext = os.path.splitext(path)
+                    path = f"{name}_{i}{ext}"
+                super().save(path, mask=mask)
```
Comment (liz-is) [inline rex_xai/multi_explanation.py:32]: I think you can revert these changes - interactive plotting is more complicated than I thought it would be, I remembered I actually added a new `show()` method for `Explanation` so I think we'd want to override that instead for `MultiExplanation`. That might need a modified form of `save_multi_explanation`, too. I'll have a look at how to do it.

**Review by liz-is — APPROVED** [review]

File: rex_xai/multi_explanation.py (hunk 8 above)
Comment (kellino) [reply rex_xai/multi_explanation.py:66]: Ah, this is a subtle point. The total number of explanations does not change, but now we include the global max explanation first, and then other explanations. This makes more sense to me (we should definitely include the "best" explanations) and is in line with what we say in the draft paper

File: rex_xai/config.py [hunk 10]
```diff
@@ -548,7 +530,7 @@ def load_config(config_path=None):
             process_config_dict(config_file_args, default_args)
             return default_args
         except Exception as e:
-            print(f"{e}, reverting to default args")
+            logger.warn("%s, reverting to default args", e)
```
[the remark this answers is not in the document]
Comment (kellino) [reply rex_xai/config.py:533]: That's a good idea. I'll add that in now

File: rex_xai/multi_explanation.py (hunk 9 above)
Comment (kellino) [reply rex_xai/multi_explanation.py:32]: will do

File: rex_xai/multi_explanation.py (hunk 8 above)
Comment (liz-is) [reply rex_xai/multi_explanation.py:63]: Ahh, got it. Then we should be careful how we document this, since `self.args.spotlights` won't actually be the number of spotlights that are used, right?
